# Multiplication (sur 8 bits)
## par addition et décalage

Principe : pour chaque ième bit de a à 1, ajouter b << i

En C cela donnerait (pour 8 bits) :

<pre>
unsigned int mult (unsigned int a, unsigned int b) {
   unsigned int resultat=0; 
   while (a != 0) {
     if ((a & 1) != 0) {// ajouter b si a_0 == 1
        resultat = resultat + b;}
     b = b << 1; a = a >> 1;}  
   return resultat&255; }
</pre>

En assembleur, cela donne (sur 8 bits) :   

In [34]:
   MOVS  R0, #11 @ a
   MOVS  R1, #13 @ b
   MOVS  R2, #0  @ resultat
   MOVS  R3, #1 @cst 1
   MOVS  R4, #0 @cst 0
   MOVS  R5, #255 @cst 255 (masque pour 1 octet)
   BAL   ct 
tq TST   R0, R3
   BEQ   st
   ADDS   R2, R2, R1  
st LSLS   R1, R1, #1
   LSRS   R0, R0, #1
ct CMP   R0, R4
   BNE   tq
   ANDS   R2, R2, R5 @ masque sur 8 bits 

In [36]:
%run 1000
%reg R2

R2: 143


Avec R0=11 et R1=13 vous devriez obtenir R2=143

Avec R0=245 et R1=13 vous devriez obtenir le produit de -11 par 13, car sur 8 bits 245=(256)-11, donc 245 c'est comme -11. Revenons au produit de -11 par 13 (ou 245 par 13), vous devriez obtenir R2=113, car 113=(256)-143, donc obtenir 113, c'est comme obtenir -143 (c'est à dire -11*13 !).
